src/librustc_mir/transform/const_prop.rs

```diff
@@ -312,21 +315,21 @@ impl<'tcx> Visitor<'tcx> for CanConstProp {
         match statement.kind {
             StatementKind::SetDiscriminant { ref place, .. } |
             StatementKind::Assign(ref place, _) => {
-                if self.is_our_local(place) {
-                    if self.found_assignment {
-                        self.can_const_prop = false;
+                if let Some(local) = place_to_local(place) {
```

Comment:
This should be done with `visit_local` and `PlaceContext` - see  https://github.com/rust-lang/rust/blob/fe7e1a45f37f4265434cead827f587e75412f85c/src/librustc_mir/transform/promote_consts.rs#L79-L135